# Preprocessing

In [52]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
df = pd.read_csv("data/vgsales.csv")
df

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27      8.89     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00      0.01      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

### それぞれのカラムの欠損値の数を確認

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


データの数である16598に対して、 YearカラムとPublisherカラムの Non-Null のデータカウントが足りていないことがわかる。  
**つまり、YearカラムとPublisherカラムには欠損値があることがわかる。**

#### 欠損値の数

In [54]:
print("Publisher カラムの欠損値の数:",len(df[df["Publisher"].isna()]))
print("Year カラムの欠損値の数:",len(df[df["Year"].isna()]))

Publisher カラムの欠損値の数: 58
Year カラムの欠損値の数: 271


## 欠損値に他の値を代入

In [55]:
# pub_na_idx = df[df["Publisher"].isna()].index
# year_na_idx = df[df["Year"].isna()].index

# # 欠損値を "NaN" という文字列で置き換える
# df[["Publisher"]] = df[["Publisher"]].fillna("NaN")
# df[["Year"]] = df[["Year"]].fillna(df["Year"].median())
# df.iloc[pub_na_idx]
# df.iloc[year_na_idx]

# 欠損値の置き換えは以下のように, inplace=True オプションを指定することで、1行でできる
df.fillna({"Publisher": "NaN", "Year": df["Year"].median()}, inplace=True)

`df[Publisher] == "unknown"`となる行も、`isnaメソッド`では引っかからないが欠損値なので対応する  
Publisherカラムのうち欠損値であった物だけでDataFrameを作っておく

In [56]:
pub_na_df = df[df["Publisher"] == "NaN"]
pub_unkown_df = df[df["Publisher"]=="Unknown"]
pub_missing_df = pd.concat([pub_na_df, pub_unkown_df])
pub_missing_df

Rank                                               Name Platform  \
470      471                         wwe Smackdown vs. Raw 2006      PS2   
1303    1305                                     Triple Play 99       PS   
1662    1664       Shrek / Shrek 2 2-in-1 Gameboy Advance Video      GBA   
2222    2224                                 Bentley's Hackpack      GBA   
3159    3161  Nicktoons Collection: Game Boy Advance Video V...      GBA   
...      ...                                                ...      ...   
16460  16463     Horse Life 4: My Horse, My Friend, My Champion      3DS   
16493  16496                                         The Smurfs      3DS   
16532  16535                            STORM: Frontline Nation       PC   
16545  16548  The Treasures of Mystery Island 3 Pack - Save ...       PC   
16581  16584                                          Fit & Fun      Wii   

         Year     Genre Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
470    2007.0  Fighting       NaN      1.57      1.02       0.0         0.41   
1303   2007.0    Sports       NaN      0.81      0.55       0.0         0.10   
1662   2007.0      Misc       NaN      0.87      0.32       0.0         0.02   
2222   2005.0      Misc       NaN      0.67      0.25       0.0         0.02   
3159   2004.0      Misc       NaN      0.46      0.17       0.0         0.01   
...       ...       ...       ...       ...       ...       ...          ...   
16460  2015.0    Action   Unknown      0.00      0.01       0.0         0.00   
16493  2007.0    Action   Unknown      0.00      0.01       0.0         0.00   
16532  2011.0  Strategy   Unknown      0.00      0.01       0.0         0.00   
16545  2011.0    Puzzle   Unknown      0.01      0.00       0.0         0.00   
16581  2011.0    Sports   Unknown      0.00      0.01       0.0         0.00   

       Global_Sales  
470            3.00  
1303           1.46  
1662           1.21  
2222           0.93  
3159           0.64  
...             ...  
16460          0.01  
16493          0.01  
16532          0.01  
16545          0.01  
16581          0.01  

[261 rows x 11 columns]

Publisherカラムにおける、`NaN`と`Unknown`のデータの分布を確認する

In [57]:
sns.pairplot(pub_missing_df, hue="Publisher")

上の結果より、NaNのデータは特定のYearに集中しているように見える。
=> NaNとUnkwonを別で扱っても良いかも

### sklearnのSimpleInputerを用いて欠損値の代入を行う
※ 訓練データとテストデータで分ける場合は学習データのみ欠損値を代入することに注意

In [62]:
from sklearn.impute import SimpleImputer
df = pd.read_csv("data/vgsales.csv")
pub_na_idx = df[df["Publisher"].isna()].index
year_na_idx = df[df["Year"].isna()].index

imputer = SimpleImputer(strategy="median")
df["Year"] = imputer.fit_transform(df[["Year"]])

imputer = SimpleImputer(strategy="most_frequent")
df["Publisher"] = imputer.fit_transform(df[["Publisher"]])

Yearカラムに中央値を入れると、PlatformがGBA(ゲームボーイアドバンス)の場合でも中央値の2007年が入り、明らかにおかしい欠損値の扱いになっている

**=> Platformごとにある程度現実性のあるYearの値で欠損値を補う**  
### 各PlatformでYearの中央値を計算して欠損値補完する

In [76]:
df = pd.read_csv("data/vgsales.csv")
year_na_idx = df[df["Year"].isna()].index

platform_year_dict = df.groupby("Platform").median()["Year"].to_dict()
df["Year"] = df.apply(
    lambda row: platform_year_dict[row["Platform"]] if (np.isnan(row["Year"]) and row["Platform"] in platform_year_dict) else row["Year"],
    axis=1
)
df.iloc[year_na_idx]      


Rank                         Name Platform    Year         Genre  \
179      180              Madden NFL 2004      PS2  2005.0        Sports   
377      378             FIFA Soccer 2004      PS2  2005.0        Sports   
431      432   LEGO Batman: The Videogame      Wii  2009.0        Action   
470      471   wwe Smackdown vs. Raw 2006      PS2  2005.0      Fighting   
607      608               Space Invaders     2600  1982.0       Shooter   
...      ...                          ...      ...     ...           ...   
16307  16310                Freaky Flyers       GC  2003.0        Racing   
16327  16330                    Inversion       PC  2009.0       Shooter   
16366  16369  Hakuouki: Shinsengumi Kitan      PS3  2011.0     Adventure   
16427  16430                 Virtua Quest       GC  2003.0  Role-Playing   
16493  16496                   The Smurfs      3DS  2013.0        Action   

                                    Publisher  NA_Sales  EU_Sales  JP_Sales  \
179                           Electronic Arts      4.26      0.26      0.01   
377                           Electronic Arts      0.59      2.36      0.04   
431    Warner Bros. Interactive Entertainment      1.86      1.02      0.00   
470                                       NaN      1.57      1.02      0.00   
607                                     Atari      2.36      0.14      0.00   
...                                       ...       ...       ...       ...   
16307                                 Unknown      0.01      0.00      0.00   
16327                      Namco Bandai Games      0.01      0.00      0.00   
16366                                 Unknown      0.01      0.00      0.00   
16427                                 Unknown      0.01      0.00      0.00   
16493                                 Unknown      0.00      0.01      0.00   

       Other_Sales  Global_Sales  
179           0.71          5.23  
377           0.51          3.49  
431           0.29          3.17  
470           0.41          3.00  
607           0.03          2.53  
...            ...           ...  
16307         0.00          0.01  
16327         0.00          0.01  
16366         0.00          0.01  
16427         0.00          0.01  
16493         0.00          0.01  

[271 rows x 11 columns]

## kNNを用いて欠損値代入
※ 欠損値を代表値で代入する場合は、交差検証(Cross Validation)やhold-outで実施する必要がある(学習データとテストデータを分ける必要がある)

In [98]:
from sklearn.preprocessing import StandardScaler

df = pd.read_csv("data/vgsales.csv")
df[["Publisher"]] = df[["Publisher"]].fillna("NaN")
# 距離を考えられないカラムを抜いておく
X = df.drop("Name", axis=1, inplace=True)
target = "Year"
# targetであるX以外の特徴量でkNNを用いてtargetの特徴量を予測するので、Xをdropしておく
X = df.drop(target, axis=1)
y = df[target]
# 数値型のカラムを抽出
num_cols = X.select_dtypes(include=np.number).columns.to_list()
# ダミー変数
X = pd.get_dummies(X, drop_first=True)
# 数値型のカラムを標準化する
X[num_cols] = StandardScaler().fit_transform(X[num_cols])

In [99]:
# 欠損値以外を学習データとする。欠損値は予測されるデータ。
test_indexes = df[df["Year"].isna()].index
train_indexes = df[~df["Year"].isna()].index

X_train, X_test = X.iloc[train_indexes], X.iloc[test_indexes]
y_train, y_test = y.iloc[train_indexes], y.iloc[test_indexes]
# 予測前の y_test は全てNaNとなっている。

In [100]:
from sklearn.neighbors import KNeighborsRegressor
knn = KNeighborsRegressor(n_neighbors=3).fit(X_train, y_train)
y_pred = knn.predict(X_test)

#### kNNImputerを用いた方法

In [119]:
from sklearn.impute import KNNImputer
df = pd.get_dummies(df, drop_first=True)
# 数値型のカラムを標準化する
df[num_cols] = StandardScaler().fit_transform(df[num_cols])

imputer = KNNImputer(n_neighbors=3)
imputer.set_output(transform="pandas")
df_imputed = imputer.fit_transform(df)

In [120]:
df_imputed.iloc[test_indexes]

Rank         Year  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
179   -1.694720  2004.333333  4.892293  0.224302 -0.219159     3.510062   
377   -1.653398  2005.000000  0.398370  4.379953 -0.122160     2.449519   
431   -1.642129  2008.666667  1.953488  1.728252 -0.251492     1.282922   
470   -1.633990  2003.333333  1.598383  1.728252 -0.251492     1.919248   
607   -1.605399  1980.666667  2.565739 -0.013164 -0.251492    -0.095783   
...         ...          ...       ...       ...       ...          ...   
16307  1.671511  2003.333333 -0.311841 -0.290207 -0.251492    -0.254864   
16327  1.675685  2009.666667 -0.311841 -0.290207 -0.251492    -0.254864   
16366  1.683824  2014.000000 -0.311841 -0.290207 -0.251492    -0.254864   
16427  1.696554  2009.333333 -0.311841 -0.290207 -0.251492    -0.254864   
16493  1.710328  2013.333333 -0.324086 -0.270418 -0.251492    -0.254864   

       Global_Sales  Platform_3DO  Platform_3DS  Platform_DC  ...  \
179        3.017760           0.0           0.0          0.0  ...   
377        1.898775           0.0           0.0          0.0  ...   
431        1.692985           0.0           0.0          0.0  ...   
470        1.583659           0.0           0.0          0.0  ...   
607        1.281404           0.0           0.0          0.0  ...   
...             ...           ...           ...          ...  ...   
16307     -0.339194           0.0           0.0          0.0  ...   
16327     -0.339194           0.0           0.0          0.0  ...   
16366     -0.339194           0.0           0.0          0.0  ...   
16427     -0.339194           0.0           0.0          0.0  ...   
16493     -0.339194           0.0           1.0          0.0  ...   

       Publisher_Zushi Games  Publisher_bitComposer Games  \
179                      0.0                          0.0   
377                      0.0                          0.0   
431                      0.0                          0.0   
470                      0.0                          0.0   
607                      0.0                          0.0   
...                      ...                          ...   
16307                    0.0                          0.0   
16327                    0.0                          0.0   
16366                    0.0                          0.0   
16427                    0.0                          0.0   
16493                    0.0                          0.0   

       Publisher_dramatic create  Publisher_fonfun  Publisher_iWin  \
179                          0.0               0.0             0.0   
377                          0.0               0.0             0.0   
431                          0.0               0.0             0.0   
470                          0.0               0.0             0.0   
607                          0.0               0.0             0.0   
...                          ...               ...             ...   
16307                        0.0               0.0             0.0   
16327                        0.0               0.0             0.0   
16366                        0.0               0.0             0.0   
16427                        0.0               0.0             0.0   
16493                        0.0               0.0             0.0   

       Publisher_id Software  Publisher_imageepoch Inc.  \
179                      0.0                        0.0   
377                      0.0                        0.0   
431                      0.0                        0.0   
470                      0.0                        0.0   
607                      0.0                        0.0   
...                      ...                        ...   
16307                    0.0                        0.0   
16327                    0.0                        0.0   
16366                    0.0                        0.0   
16427                    0.0                        0.0   
16493                    0.0                        0.0   

自分で実装した結果と kNNImputer を用いた結果を比較する

In [121]:
y_pred

array([2004.33333333, 2005.        , 2008.66666667, 2003.33333333,
       1980.66666667, 2009.33333333, 1990.66666667, 2009.        ,
       2009.66666667, 2009.        , 2010.66666667, 2009.66666667,
       1999.        , 2010.66666667, 2008.        , 1982.33333333,
       1982.        , 2002.33333333, 2003.33333333, 2008.        ,
       1998.66666667, 1997.66666667, 2013.66666667, 1982.        ,
       2005.        , 2010.66666667, 2014.33333333, 2002.        ,
       2001.        , 2009.66666667, 2014.33333333, 2014.66666667,
       2006.66666667, 2003.66666667, 2010.33333333, 2009.        ,
       2009.33333333, 2001.66666667, 2011.33333333, 2008.        ,
       2004.        , 2011.        , 2006.66666667, 2006.        ,
       2002.33333333, 2009.        , 2007.33333333, 2003.33333333,
       2004.33333333, 2004.33333333, 2003.        , 1983.66666667,
       2012.33333333, 2010.        , 1983.33333333, 2009.        ,
       1982.33333333, 2003.33333333, 1982.33333333, 2011.66666